src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+        }
+    }
+
+    private class FilteringIntersection extends Intersection
+    {
+        private final IntersectVisitor visitor;
+        private final byte[] packedValue;
+        private final short[] origIndex;
+
+        FilteringIntersection(IndexInput treeInput, IndexInput postingsInput, IndexInput postingsSummaryInput,
+                              PackedIndexTree index, IntersectVisitor visitor,
+                              QueryEventListener.BalancedTreeEventListener listener, QueryContext context)
+        {
+            super(treeInput, postingsInput, postingsSummaryInput, index, listener, context);
+            this.visitor = visitor;
+            this.packedValue = new byte[packedBytesLength];
+            this.origIndex = new short[maxPointsInLeafNode];
+        }
+
+        @Override
+        public void executeInternal(final PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            collectPostingLists(postingLists, minPackedValue, maxPackedValue);
+        }
+
+        private void collectPostingLists(PriorityQueue<PeekablePostingList> postingLists, byte[] cellMinPacked, byte[] cellMaxPacked) throws IOException
+        {
+            context.checkpoint();
+
+            final Relation r = visitor.compare(cellMinPacked, cellMaxPacked);
+
+            if (r == Relation.CELL_OUTSIDE_QUERY)
+            {
+                // This cell is fully outside the query shape: stop recursing
+                return;
+            }
+
+            if (r == Relation.CELL_INSIDE_QUERY)
+            {
+                // This cell is fully inside the query shape: recursively add all points in this cell without filtering
+                super.collectPostingLists(postingLists);
+                return;
+            }
+
+            if (index.isLeafNode())
+            {
+                if (index.nodeExists())
+                    filterLeaf(postingLists);
+                return;
+            }
+
+            visitNode(postingLists, cellMinPacked, cellMaxPacked);
+        }
+
+        private void filterLeaf(PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            treeInput.seek(index.getLeafBlockFP());
+
+            int count = treeInput.readVInt();
+
+            // loading doc ids occurred here prior
+
+            int orderMapLength = treeInput.readVInt();
+
+            long orderMapPointer = treeInput.getFilePointer();
+
+            SeekingRandomAccessInput randomAccessInput = new SeekingRandomAccessInput(treeInput);
+            for (int x = 0; x < count; x++)
+            {
+                origIndex[x] = (short) LeafOrderMap.getValue(randomAccessInput, orderMapPointer, x, leafOrderMapReader);
+            }
+
+            // seek beyond the ordermap
+            treeInput.seek(orderMapPointer + orderMapLength);
+
+            FixedBitSet fixedBitSet = visitDocValues(treeInput, count, visitor, origIndex);
```

Comment:
So to confirm my understanding, the bitset here represents the indexes of the leaf postings. At least from the usage in `FilteringPostingList`, it seems that way...

```
int position = 0;
...
if (filter.get(position++))
{
    return segmentRowId;
}
```

Reply:
That is my understanding. This area of the code is overly confusing because we have an `OrdinalPostingList` that returns the position of the posting in the posting list rather than the posting itself and then this is used to check the bitset. If the bitset says ok then we return the row ID matching the position from the posting list.